### ISSUE TITLE
Driver swallows exception when calling stored procedure with OUT param

### ISSUE THREAD

**Issue by jon32446** [microsoft/mssql-jdbc#2190]
## Driver version
mssql-jdbc-12.4.0.jre8.jar

## SQL Server version
Microsoft SQL Server 2019 (RTM-GDR) (KB5014356) - 15.0.2095.3 (X64) 
	Apr 29 2022 18:00:13 
	Copyright (C) 2019 Microsoft Corporation
	Developer Edition (64-bit) on Windows Server 2019 Standard 10.0 <X64> (Build 17763: ) (Hypervisor)

## Client Operating System
Windows Server 2019 Standard

## JAVA/JVM version
Java 17
```
>javac -version
javac 17

>java -version
openjdk version "17" 2021-09-14 LTS
OpenJDK Runtime Environment Zulu17.28+13-CA (build 17+35-LTS)
OpenJDK 64-Bit Server VM Zulu17.28+13-CA (build 17+35-LTS, mixed mode, sharing)

>where java
C:\Program Files\Zulu\zulu-17\bin\java.exe
```

## Table schema
No tables needed, but the following stored procedure is needed.
```
IF OBJECT_ID('test123', 'P') IS NOT NULL DROP PROCEDURE test123
GO
CREATE PROCEDURE test123
	@out INT OUT
AS BEGIN
	SET @out = 1
	SELECT TOP 0 * FROM sys.all_objects
	RAISERROR('Test error', 16, 1)
END
```

## Sample code to reproduce issue
```
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.CallableStatement;
import java.sql.Statement;
import java.sql.ResultSet;
import java.sql.DriverManager;
import java.sql.Types;

public class Test123
{
	private static String connectionUrl = "jdbc:sqlserver://localhost:1433;" +
		"databaseName=test;username=sa;password=test;encrypt=false;";
	
	public static void main(String[] args) throws Exception
	{
		testCallableStatement(true); //     out param getInt - no error - wrong
		testCallableStatement(false); // no out param getInt - has error
		/*
		 * >javac Test123.java && java -cp ".;.\mssql-jdbc-12.4.0.jre8.jar" Test123
		 * Exception in thread "main" com.microsoft.sqlserver.jdbc.SQLServerException: Test error
			  at com.microsoft.sqlserver.jdbc.SQLServerException.makeFromDatabaseError(SQLServerException.java:255)
			  at com.microsoft.sqlserver.jdbc.SQLServerStatement.getNextResult(SQLServerStatement.java:1695)
			  at com.microsoft.sqlserver.jdbc.SQLServerStatement.getMoreResults(SQLServerStatement.java:1406)
			  at Test123.testCallableStatement(Test123.java:28)
			  at Test123.main(Test123.java:17)
		 */
	}
	
	public static void testCallableStatement(boolean getOutParam) throws Exception
	{
		Connection cn = DriverManager.getConnection(connectionUrl);
		CallableStatement stmt = cn.prepareCall("{call test123(?)}");
		stmt.registerOutParameter(1, Types.INTEGER);
		ResultSet rs = stmt.executeQuery();
		if (getOutParam)
			stmt.getInt(1);
		while (stmt.getMoreResults())
		{
			stmt.getResultSet();
			if (stmt.getUpdateCount() == -1)
			{
				break;
			}
		}
	}
}
```

## Problem description
The driver swallows an error inside a stored procedure if there is a result set returned by the proc and an OUT parameter that gets read by the Java.

## Expected behavior
When a stored procedure is called, any exception that occurs in the stored procedure should result in a SQLException in the Java.

## Actual behavior
When calling a stored procedure that:

* returns a result set
* has an OUT parameter
* raises an error after the result set

then no SQLException is raised in the Java.

## Error message/stack trace
Please note the Java repro code above. This exception only happens when the OUT param is *not* read.
```
Exception in thread "main" com.microsoft.sqlserver.jdbc.SQLServerException: Test error
        at com.microsoft.sqlserver.jdbc.SQLServerException.makeFromDatabaseError(SQLServerException.java:255)
        at com.microsoft.sqlserver.jdbc.SQLServerStatement.getNextResult(SQLServerStatement.java:1695)
        at com.microsoft.sqlserver.jdbc.SQLServerStatement.getMoreResults(SQLServerStatement.java:1406)
        at Test123.testCallableStatement(Test123.java:28)
        at Test123.main(Test123.java:17)
```

## Any other details that can be helpful
If the stored procedure has an OUT param but does not return a result set before the error, then the SQLException is raised as expected.

## JDBC trace logs
I couldn't get the tracing to work.

**Comment by Jeffery-Wasty** [comment]
Hi @jon32446,

Thank you for the information and repro code, we'll take a look into this issue.

**Comment by Jeffery-Wasty** [comment]
Just an update as its been a while. We haven't been able to look into this issue long enough to determine cause. We will update again when we have more information to share.

**Comment by lilgreenbird** [comment]
hi @jon32446 

The issue is that you need to process the result sets to retrieve the exception that was generated. When `getOutParam` is true and `getInt(1)` is executed it does not return a result set so `getMoreResults()` is false and no exception is retrieved. If you want to retrieve the exception from the stored procedure, you will need to process the result set to retrieve the exception right after it is executed, and then you can call getInt(1) to retrieve the param after. 

i.e.
```
	public static void testCallableStatement(boolean getOutParam) throws Exception
	{
		Connection cn = DriverManager.getConnection(connectionUrl);
		CallableStatement stmt = cn.prepareCall("{call test123(?)}");
		stmt.registerOutParameter(1, Types.INTEGER);
		ResultSet rs = stmt.executeQuery();
		while (stmt.getMoreResults())
		{
			stmt.getResultSet();
			if (stmt.getUpdateCount() == -1)
			{
				break;
			}
		}

		if (getOutParam)
			stmt.getInt(1);

	}
```

We have a [wiki ](https://github.com/microsoft/mssql-jdbc/wiki/Handling-SQLExceptions) that discusses exception handling in the driver for your reference.

**Comment by jon32446** [comment]
Hi @lilgreenbird Thanks for checking this out.

It seems like this is still a bug? I am following the wiki guidance of processing all of the result sets. It's just that I do that *after* calling getInt. It isn't obvious why the getInt has to be called after the result set is processed (and the linked wiki doesn't exclude this use case). The driver currently supports both ways in the happy day but not when there is an error.

It may not always be possible to call getInt after iterating the result set. Would the use case not be supported where the Java logic needs the value from the getInt before it iterates the result set? For example (pseudocode)
```
@mode INT OUT
SELECT @mode = mode FROM config
IF @mode = 1
	SELECT 'one kind of data'
ELSE
	SELECT 'something else'
END
```
And then the Java might have two different branches handling the result set differently based on the `@mode` param.

Not super important, but to give some context. What happened in my case is that a proc that previously returned an OUT param, but no rows, was changed to return some rows. The code that dealt with the result sets was added after the OUT param was read. Everything worked fine in the happy day cases. The issue with errors being swallowed was luckily caught by an older regression test case and a workaround was implemented (we got rid of the result sets). In our case, we could move the getInt after the result set iteration, but it does seem like the driver should support the getInt being called both before and after result set iteration.
